dll/shellext/fontext/CFontExt.cpp

```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
     return S_OK;
 }
 
+// *** IDropTarget methods ***
+STDMETHODIMP CFontExt::DragEnter(IDataObject* pDataObj, DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    FORMATETC formatETC;
+    HRESULT hr = DoGetFormatETC(pDataObj, formatETC);
+    if (hr != S_OK)
+    {
+        *pdwEffect = DROPEFFECT_NONE;
+        return DRAGDROP_S_CANCEL;
+    }
+
+    // TODO: Check filename extension
+
+#if 1   // Please implement DoGetFontTitle
+    *pdwEffect = DROPEFFECT_NONE;
+    return DRAGDROP_S_CANCEL;
+#else
+    *pdwEffect = DROPEFFECT_COPY;
+    return S_OK;
+#endif
+}
+
+STDMETHODIMP CFontExt::DragOver(DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    return S_OK;
+}
+
+STDMETHODIMP CFontExt::DragLeave()
+{
+    return S_OK;
+}
+
+STDMETHODIMP CFontExt::Drop(IDataObject* pDataObj, DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    FORMATETC formatETC;
+    HRESULT hr = DoGetFormatETC(pDataObj, formatETC);
+    if (hr != S_OK)
+    {
+        *pdwEffect = DROPEFFECT_NONE;
+        return DRAGDROP_S_CANCEL;
+    }
+
+    STGMEDIUM medium;
+    hr = pDataObj->GetData(&formatETC, &medium);
+    if (FAILED(hr))
+    {
+        *pdwEffect = DROPEFFECT_NONE;
+        return DRAGDROP_S_CANCEL;
+    }
+
+    BOOL bOK = TRUE;
+    CAtlArray<CStringW> FontPaths;
+    HDROP hDrop = (HDROP)GlobalLock(medium.hGlobal);
+    if (hDrop)
+    {
+        WCHAR szPath[MAX_PATH];
+        UINT iFile, cFiles = DragQueryFile(hDrop, 0xFFFFFFFF, NULL, 0);
+        for (iFile = 0; iFile < cFiles; ++iFile)
+        {
+            DragQueryFile(hDrop, iFile, szPath, _countof(szPath));
+
+            if (PathIsDirectoryW(szPath))
+            {
+                bOK = FALSE;
+                break;
+            }
+
+            LPCWSTR pchDotExt = PathFindExtensionW(szPath);
+            if (_wcsicmp(pchDotExt, L".ttf") != 0 &&
+                _wcsicmp(pchDotExt, L".ttc") != 0 &&
+                _wcsicmp(pchDotExt, L".otf") != 0 &&
+                _wcsicmp(pchDotExt, L".otc") != 0 &&
+                _wcsicmp(pchDotExt, L".fon") != 0 &&
+                _wcsicmp(pchDotExt, L".fnt") != 0)
+            {
+                bOK = FALSE;
+                break;
+            }
+
+            FontPaths.Add(szPath);
+        }
+
+        GlobalUnlock(medium.hGlobal);
+    }
+    ReleaseStgMedium(&medium);
+
+    if (!bOK)
+        return E_FAIL;
+
+    WCHAR szFontsDir[MAX_PATH];
+    GetWindowsDirectoryW(szFontsDir, _countof(szFontsDir));
+    PathAppendW(szFontsDir, L"Fonts");
```

Comment:
Other places in this library `SHGetFolderPathW(NULL, CSIDL_FONTS, NULL, 0, Path)` is used

Reply:
OK.